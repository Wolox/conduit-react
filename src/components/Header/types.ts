import { IconDefinition } from '@fortawesome/fontawesome-svg-core';

export interface ItemsMenu {
  text: string;
  href: string;
  isProtected: boolean;
  icon: IconDefinition;
}
