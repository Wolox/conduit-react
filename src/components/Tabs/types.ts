import { ApiResponse } from 'apisauce';
import { UseQueryResult } from 'react-query';

import { Articles } from 'types/Article';

export interface Tab<T = void> {
  text: string;
  list: (data?: T) => UseQueryResult<ApiResponse<Articles>>;
}
