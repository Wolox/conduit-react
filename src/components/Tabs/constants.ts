export const CONFIG_TAB_FEED = {
  isPrivate: true,
  text: 'Tabs:yourFeed',
  endpoint: () => {
    console.log('endpoint Tabs:yourFeed');
  },
  visible: false
};

export const CONFIG_TAB_GLOBAL = {
  isPrivate: true,
  text: 'Tabs:globalFeed',
  endpoint: () => {
    console.log('endpoint Tabs:globalFeed');
  },
  visible: true
};

export const CONFIG_TAB_MY_POSTS = {
  isPrivate: true,
  text: 'Tabs:post',
  endpoint: () => {
    console.log('endpoint Tabs:post');
  },
  visible: false
};

export const CONFIG_TAB_FAVORITES = {
  isPrivate: true,
  text: 'Tabs:favorites',
  endpoint: () => {
    console.log('endpoint Tabs:favorites');
  },
  visible: false
};
