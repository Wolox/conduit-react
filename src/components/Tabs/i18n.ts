import i18next from 'i18next';

i18next.addResources('es', 'Tabs', {
  yourFeed: 'Publicaciones Seguidas',
  globalFeed: 'Todas las Públicaciones',
  post: 'Mis Públicaciones',
  favorites: 'Públicaciones Favoritas'
});

i18next.addResources('en', 'Tabs', {
  yourFeed: 'Your Feed',
  globalFeed: 'Global Feed',
  post: 'My Articles',
  favorites: 'Favorites Articles'
});
