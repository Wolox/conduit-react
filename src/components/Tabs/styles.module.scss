@import 'scss/variables/_colors';

.tabs {
  border-bottom: 1px solid $gray;

  .tab {
    color: $black;
    cursor: pointer;
    font-size: 18px;
    margin: 0 5px;
    opacity: 0.5;
    padding: 15px;

    &:hover {
      opacity: 1;
    }
  }

  .active {
    border-bottom: 2px solid $green-apple;
    color: $green-apple;
    opacity: 1;

    &:hover {
      opacity: 0.8;
    }
  }
}
