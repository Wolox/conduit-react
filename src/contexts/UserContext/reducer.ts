import { produce } from 'immer';
import { Reducer } from 'react';

import { UserProfile } from 'types/User';

export interface User {
  id: number;
  sessionToken: string;
}
export interface UserState {
  user: UserProfile | null;
}

export interface Credentials {
  username: string;
  password: string;
}

export const INITIAL_STATE: UserState = {
  user: null
};

enum ActionTypes {
  SET_USER = 'SET_USER',
  RESET_USER = 'RESET_USER',
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT'
}

interface SetUser {
  type: ActionTypes.SET_USER;
  payload: UserProfile;
}

interface ResetUser {
  type: ActionTypes.RESET_USER;
}

export type Action = SetUser | ResetUser;

export const actionCreators = {
  setUser: (user: UserProfile): SetUser => ({ type: ActionTypes.SET_USER, payload: user }),
  resetUser: (): ResetUser => ({ type: ActionTypes.RESET_USER })
};

export const reducer: Reducer<UserState, Action> = produce((draft, action) => {
  switch (action.type) {
    case ActionTypes.SET_USER: {
      draft.user = action.payload;
      break;
    }
    case ActionTypes.RESET_USER: {
      draft.user = null;
      break;
    }
    // no default
  }
}, INITIAL_STATE);
