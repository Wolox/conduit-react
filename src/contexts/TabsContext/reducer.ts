import { produce } from 'immer';
import { Reducer } from 'react';

import { Tab } from 'components/Tabs/types';
import { CONFIG_TAB_GLOBAL } from 'constants/tabs';

interface TabContext {
  hashtag: Tab | null;
  tabActive: Tab;
}

export const INITIAL_STATE: TabContext = {
  hashtag: null,
  tabActive: CONFIG_TAB_GLOBAL
};

enum ActionTypes {
  ACTIVE = 'ACTIVE',
  HASHTAG = 'HASHTAG'
}

interface ActiveTab {
  type: ActionTypes.ACTIVE;
  payload: {
    tab: Tab;
  };
}

interface SetHashtag {
  type: ActionTypes.HASHTAG;
  payload: {
    tab: Tab;
  };
}

export type Action = ActiveTab | SetHashtag;

export const actionCreators = {
  activeTab: (tab: Tab): ActiveTab => ({ type: ActionTypes.ACTIVE, payload: { tab } }),
  setHastag: (tab: Tab): SetHashtag => ({ type: ActionTypes.HASHTAG, payload: { tab } })
};

export const reducer: Reducer<TabContext, Action> = produce((draft, action) => {
  switch (action.type) {
    case ActionTypes.ACTIVE: {
      draft.tabActive = action.payload.tab;
      draft.hashtag = null;
      break;
    }
    case ActionTypes.HASHTAG: {
      draft.hashtag = action.payload.tab;
      draft.tabActive = action.payload.tab;
      break;
    }
    // no default
  }
}, INITIAL_STATE);
