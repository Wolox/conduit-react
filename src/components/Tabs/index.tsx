import { useTranslation } from 'react-i18next';
import { memo } from 'react';

import { useDispatch, useSelector } from 'contexts/TabsContext';
import { actionCreators } from 'contexts/TabsContext/reducer';

import styles from './styles.module.scss';
import { Tab } from './types';

interface Props {
  tabs: Tab[];
}

function Tabs({ tabs }: Props) {
  const { t } = useTranslation();
  const { hashtag, tabActive } = useSelector((state) => state);
  const dispatch = useDispatch();

  const handleChangeTab = (tab: Tab) => {
    dispatch(actionCreators.activeTab(tab));
  };

  return (
    <div>
      <ul className={styles.tabs}>
        {tabs.map((actualTab) => (
          <li
            key={actualTab.text}
            onClick={() => handleChangeTab(actualTab)}
            className={`${styles.tab} ${tabActive.text === actualTab.text && styles.active}`}
          >
            {t(actualTab.text)}
          </li>
        ))}
        {hashtag && (
          <li key={hashtag.text} className={`${styles.tab} ${styles.active}`}>
            {t(hashtag.text)}
          </li>
        )}
      </ul>
    </div>
  );
}

export default memo(Tabs);
