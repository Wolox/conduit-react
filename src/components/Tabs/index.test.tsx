import { ReactNode } from 'react';
import { render, screen } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import userEvent from '@testing-library/user-event';

import { StateContext } from 'contexts/TabsContext';
import { actionCreators, TabContext } from 'contexts/TabsContext/reducer';

import { Tab } from './types';

import Tabs from '.';

export const FACEBOOK_TAB: Tab = {
  text: 'FACEBOOK',
  list: jest.fn()
};

export const INSTAGRAM_TAB: Tab = {
  text: 'INSTAGRAM',
  list: jest.fn()
};

export const TWITTER_TAG: Tab = {
  text: 'TWITTER_TAG',
  list: jest.fn()
};

const INIT_TABS = [FACEBOOK_TAB, INSTAGRAM_TAB];

export const INITIAL_STATE: TabContext = {
  hashtag: TWITTER_TAG,
  tabActive: FACEBOOK_TAB
};

function Wrapped({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient();
  const history = createMemoryHistory();
  return (
    <QueryClientProvider client={queryClient}>
      <StateContext.Provider value={INITIAL_STATE}>
        <Router history={history}>{children}</Router>
      </StateContext.Provider>
    </QueryClientProvider>
  );
}

describe('Component Tabs', () => {
  test('Renders component', () => {
    const { asFragment } = render(
      <Wrapped>
        <Tabs tabs={INIT_TABS} />
      </Wrapped>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Tab active for default', () => {
    render(
      <Wrapped>
        <Tabs tabs={INIT_TABS} />
      </Wrapped>
    );
    const tabActive = screen.getByText(FACEBOOK_TAB.text);
    expect(tabActive).toHaveClass('active');
  });

  test('Add tab tag if is send', () => {
    render(
      <Wrapped>
        <Tabs tabs={INIT_TABS} />
      </Wrapped>
    );

    expect(screen.getByText(TWITTER_TAG.text)).toBeInTheDocument();
  });

  test('Change Tab active when is clicked', () => {
    render(
      <Wrapped>
        <Tabs tabs={INIT_TABS} />
      </Wrapped>
    );

    const fnDispatch = jest.fn();
    actionCreators.activeTab = fnDispatch;

    userEvent.click(screen.getByText(INSTAGRAM_TAB.text));
    expect(fnDispatch).toHaveBeenCalledWith(INSTAGRAM_TAB);
  });
});
