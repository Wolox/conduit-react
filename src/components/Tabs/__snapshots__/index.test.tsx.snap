// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Tabs Renders component 1`] = `
<DocumentFragment>
  <div>
    <ul
      class="row middle start tabs"
    >
      <li
        class="tab active"
      >
        FACEBOOK
      </li>
      <li
        class="tab"
      >
        INSTAGRAM
      </li>
      <li
        class="tab active"
      >
        TWITTER_TAG
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
